compiler_opt/es/gradient_ascent_optimization_algorithms.py

```diff
@@ -0,0 +1,183 @@
+# https://arxiv.org/abs/1804.02395
+#
+###############################################################################
+r"""Library of gradient ascent algorithms.
+
+Library of stateful gradient descent algorithms taking as input the gradient and
+current parameters, and output the new parameters.
+"""
+
+import abc
+import numpy as np
+
+
+# TODO(kchoro): Borrow JAXs optimizer library here. Integrated into Blackbox-v2.
+class GAOptimizer(metaclass=abc.ABCMeta):
+  """Abstract class for general gradient ascent optimizers.
+
+  Class is responsible for encoding different gradient ascent optimization
+  techniques.
+  """
+
+  @abc.abstractmethod
+  def run_step(self, current_input, gradient):
+    """Conducts a single step of gradient ascent optimization.
+
+    Conduct a single step of gradient ascent optimization procedure, given the
+    current parameters and the raw gradient.
+
+    Args:
+      current_input: the current parameters.
+      gradient: the raw gradient.
+
+    Returns:
+      New parameters by conducting a single step of gradient ascent.
+    """
+    raise NotImplementedError("Abstract method")
+
+  @abc.abstractmethod
+  def get_state(self):
+    """Returns the state of the optimizer.
+
+    Returns the state of the optimizer.
+
+    Args:
+
+    Returns:
+      The state of the optimizer.
+    """
+    raise NotImplementedError("Abstract method")
+
+  @abc.abstractmethod
+  def set_state(self, state):
+    """Sets up the internal state of the optimizer.
+
+    Sets up the internal state of the optimizer.
+
+    Args:
+      state: state to be set up
+
+    Returns:
+    """
+    raise NotImplementedError("Abstract method")
+
+
+class MomentumOptimizer(GAOptimizer):
+  """Class implementing momentum gradient ascent optimizer.
+
+  Setting momentum coefficient to zero is equivalent to vanilla gradient
+  ascent.
+  """
+
+  def __init__(self, step_size, momentum):
+    self.step_size = step_size
+    self.momentum = momentum
+
+    self.moving_average = np.asarray([], dtype=np.float32)
+    super().__init__()
+
+  def run_step(self, current_input, gradient):
+    if self.moving_average.size == 0:
+      # Initialize the moving average
+      self.moving_average = np.zeros(len(current_input), dtype=np.float32)
+    elif len(self.moving_average) != len(current_input):
+      raise ValueError(
+          "Dimensions of the parameters and moving average do not match")
+
+    if not isinstance(gradient, np.ndarray):
+      gradient = np.asarray(gradient, dtype=np.float32)
+
+    self.moving_average = self.momentum * self.moving_average + (
+        1 - self.momentum) * gradient
+    step = self.step_size * self.moving_average
+
+    return current_input + step
+
+  def get_state(self):
+    return self.moving_average.tolist()
+
+  def set_state(self, state):
+    self.moving_average = np.asarray(state, dtype=np.float32)
+
+
+class AdamOptimizer(GAOptimizer):
+  """Class implementing ADAM gradient ascent optimizer."""
```

Comment:
Please describe the state vector here and in other optimizers' docstrings.  Looks like here it's [mmv1] + [mmv2] + [1] where the last dim corresponds to the current step?

Reply:
Descriptions for the states have been added now